### DISCUSSION TITLE
on-demand mp3 to client -  (port : _, _, _) -> _

### DISCUSSION THREAD

**Discussion by redimongo** [savonet/liquidsoap#3632] [Q&A]
I am trying to get a single mp3 file to feel like a live stream.

I would like the liquidsoap to allow clients to connect to port 8080/listen to access the mp3

I have tried 

```
# Import the required libraries
set("log.stdout", true)
set("harbor.bind_addr", "0.0.0.0") # Bind to all network interfaces

# Define the input source to loop the audio file
audio = single("audiofile.mp3")

# Create an HTTP server with harbor
harbor = harbor.http(
    port = 8080,     # The port where your HTTP server will listen
    "/stream",       # The mount point
    audio            # The input source defined above
)

# Start the harbor
output.dummy(harbor, fallible=true)
``` 

But then I get errors

```
harbor = harbor.http(

Error 5: this value has type
  unit
but it should be a subtype of
  (port : _, _, _) -> _

```

which is not helpful.

**Comment by vitoyucepi** [comment]
Hi @redimongo,
Can you be more specific about what you are trying to do?
Do you want to serve files on demand or just the stream?

**Reply by redimongo** [reply]
Sure so basically what I am trying to do is have GCS (Google Cloud Server) server files that it hosts and stream them as if they are live streams. 

The listener would connect at http://localhost:8080/(Station_Name)/(Show_name)

It would fetch the latest episode mp3 from the server and serve it. 

But before I can build that whole system. I just need to know how to stream a static file via liquidsoap. 

The I can tackle the harder connecting everything. 

I have done it in Python, but the issue is the buffering it doesn't act as a live stream. But I have it buffered down to a few seconds.

**Reply by vitoyucepi** [reply]
Well, there are several ways to do this:
1. Use icecast as a streaming server. It's pretty simple, but has some drawbacks.
1. Use an http server and serve an hls stream. This method requires a player that supports hls.
1. Use a third party streaming service to deliver the content.

Let's say you want to use icecast.
First you should configure and start the icecast server.
After that you can use the script, which will look similar to this one, to send the audio stream from liquidsoap to icecast.
```ruby
audio = single("audiofile.mp3")
output.icecast(%mp3, audio, host="icecast", port=8080, mount="/live", user="source", password="hackme")
```

I should probably mention ready-to-use combines like azuracast, which provides everything you need to set up a complete streaming solution.

**Reply by redimongo** [reply]
@vitoyucepi yes I understand that there are software already, however we are using Google Cloud Storage and MongoDB. it's a bigger project, but we hit a small issue with Myriad cloud software and we are trying to get our system working with that, and for that they need it as a live http stream.

**Reply by vitoyucepi** [reply]
> for that they need it as a live http stream.

HTTP stream could use
1. icecast protocol
1. hls protocol
1. http+range

It's possible to use liquidsoap's internal server, but it has performance problems at high load.
If you want a high performance solution, you should either use
1. icecast as stream distribution server
1. hls and google cloud storage

You may find some of the information you need in the presentation by [radiofrance at fosdem2020](https://archive.fosdem.org/2020/schedule/event/om_audio_streaming/).
